### ISSUE TITLE
Dune watch mode takes the lock during its complete lifetime

### ISSUE THREAD

**Issue by picdc** [ocaml/dune#6622]
In my current setup, I run ocaml-lsp alongside `dune build @check -w`, and only build when is necessary using `dune build` (or even `dune exec`). However since 3.6.0, `dune -w` locks the `_build` directory which breaks this workflow, since `dune -w` creates a lock file and keeps it until being killed.

I'm aware that `dune build -w` would produce the executables, but during development typechecking only is significantly faster on large projects.

## Expected Behavior

`dune -w` takes the lock only during compilation, and releases it once it's done.

## Actual Behavior

`dune -w`  takes the lock during its entire lifetime, any `dune build` or `dune exec` results in 
```
Error: A running dune (pid: 2431924) instance has locked the build directory.
If this is not the case, please delete _build/.lock
```

## Reproduction

On any dune library or executable:
1. run `dune build -w @check` (only typechecking, then no library/executable can be produced)
2. in another terminal, `dune build`

## Specifications

- Version of `dune` (output of `dune --version`): 3.6.1
- Version of `ocaml` (output of `ocamlc --version`): 4.14
- Operating system (distribution and version): Fedora 36

Thank you!

**Comment by Alizter** [comment]
Yes, I have wanted something similar. The correct fix would be for a new Dune build to do an RPC request to the existing Dune process in watch mode. This can be implemented today however there is a key part missing which is the relaying of error messages back from the RPC server. No technical limitation AFAIK, just a matter of doing so.
